File: examples/AODV/AODVrcvcode.m
function [exectime, data] = AODVrcvcode(seg, data)

global routing_table
global seqNbrs
global AODVparams
global verbose

myID = data.nodeID;

switch seg,
  
 case 1,
  msg = ttGetMsg;
  noww = ttCurrentTime;
  
  if (isfield(msg, 'data'))
    % Data message
    
    if (msg.dest==myID) % Data message arrived at destination
      
      if verbose
	disp(['Time: ' num2str(noww) ' Data message arrived at Node#' num2str(myID)]);
      end
      
      % Find reverse route to original source
      prev_entry = findEntry(msg.src, myID);
      
      % Update expiry timer
      prev_entry.exptime = ttCurrentTime + AODVparams.ACTIVE_ROUTE_TIMEOUT;
      updateEntry(prev_entry, myID);
      
      % Notify application
      ttTryPost('AODVRcvBox', msg.data);
      ttCreateJob('RcvTask');
      
      etime = 0.0001;
      
    else % Forward data message
      
      if verbose
	disp(['Time: ' num2str(noww) ' Node#' num2str(myID) ' about to forward data to Node#' num2str(msg.dest) ' Data: ' num2str(msg.data) ' Size: ' num2str(msg.size)]);
      end
      
      % Find reverse route to original source
      prev_entry = findEntry(msg.src, myID);
      
      % Update expiry timer
      prev_entry.exptime = ttCurrentTime + AODVparams.ACTIVE_ROUTE_TIMEOUT;
      updateEntry(prev_entry, myID);
      
      % Find next hop on route to destination
      dest_entry = findEntry(msg.dest, myID);
      
      % Update expiry timer
      dest_entry.exptime = ttCurrentTime + AODVparams.ACTIVE_ROUTE_TIMEOUT;
      updateEntry(dest_entry, myID);
      
      % Forward data message to next hop
      msg.intermed = myID;
      ttSendMsg(dest_entry.nextHop, msg, msg.size + 4);
      
      etime = 0.0001;
      
    end
    
  else % AODV control message (RREQ, RREP, RERR, or HELLO)
    
    if verbose
      disp(['Time: ' num2str(noww) ' Node#' num2str(myID) ' processing AODV message type: ' num2str(msg.type) ' from Node#' num2str(msg.intermed)]); 
    end
    
    switch msg.type,
     case 1, % RREQ
      
      if (msg.src == myID) % Skip RREQs received by the original source	
	
	etime = 0.00001; 
	
      else 
	
	% Have this RREQ already been processed?
	cache_RREQ  = findInCache(msg.src, msg.RREQID, data.cache);
	
	drop = 0;
	if (~isempty(cache_RREQ))
	  % Found in cache, drop redundant RREQ
	  
	  if verbose
	    disp(['Time: ' num2str(noww) ' Node#' num2str(myID) ' dropping redundant RREQ from Node#' num2str(msg.intermed)]);
	  end
	  drop = 1;
	end
	
	if (drop==0) % process RREQ
	    
	  if verbose
	    disp(['Time: ' num2str(noww) ' Node#' num2str(myID) ' caching RREQ with Src: ' num2str(msg.src) ' RREQID: ' num2str(msg.RREQID)]);
	  end
	  
	  % Enter RREQ in cache
	  cache_entry.src = msg.src;
	  cache_entry.RREQID = msg.RREQID;
	  l = length(data.cache);
	  data.cache{l+1} = cache_entry;
	  
	  % Create or update route entry to source
	  src_entry = findEntry(msg.src, myID);
	  
	  if (isempty(src_entry))
	    % No entry exists, create new
	    rev_entry.dest = msg.src;
	    rev_entry.nextHop = msg.intermed;
	    rev_entry.hops = msg.hopCnt + 1;
	    rev_entry.destSeqNbr = msg.srcSeqNbr;
	    rev_entry.exptime = ttCurrentTime + AODVparams.ACTIVE_ROUTE_TIMEOUT;
	    rev_entry.nbor = [];
	    rev_entry.state = 'valid';
	    
	    l = length(routing_table{myID});
	    routing_table{myID}{l+1} = rev_entry;
	    
	    % Update expiry timer	    
	    updateExpiryTimer(myID);
	    	    
	  else
	    % Update existing entry
	    src_entry.exptime = ttCurrentTime + AODVparams.ACTIVE_ROUTE_TIMEOUT;
	    src_entry.destSeqNbr = max(src_entry.destSeqNbr, msg.srcSeqNbr);
	    src_entry.nextHop = msg.intermed;
	    src_entry.hops = msg.hopCnt + 1;
	    src_entry.state = 'valid';
	    updateEntry(src_entry, myID);
	  end
	  
	  % Check if we have a route to destination
	  dest_entry = findEntry(msg.dest, myID);
	  
	  haveroute = 0;
	  if (~isempty(dest_entry))
	    haveroute = strcmp(dest_entry.state, 'valid');
	  end
	  
	  if (msg.dest==myID | haveroute)
	    % We are the destination or we have a route to it
	    
	    if verbose
	      disp(['Node#' num2str(myID) ' has a route to destination#' num2str(msg.dest)]);
	    end
	    
	    if (~isempty(dest_entry))
	      % I am not the destination, but have a route to it
	      
	      if verbose
		disp('Sending first RREP from node with route');
	      end
	      
	      dest = dest_entry.dest;
	      seqNbr = dest_entry.destSeqNbr;
	      hopCnt = dest_entry.hops;
	      lifetime = dest_entry.exptime - ttCurrentTime;
	      dest_entry.nbor = [dest_entry.nbor msg.intermed];
	      updateEntry(dest_entry, myID);
	    else
	      % I am the destination itself
	      
	      if verbose
		disp('Sending first RREP from destination itself');
	      end
	      
	      dest = myID;
	      mySeqNbr = seqNbrs(myID, myID);
	      if (mySeqNbr + 1 == msg.srcSeqNbr)
		seqNbrs(myID, myID) = seqNbrs(myID, myID) + 1;
	      end
	      seqNbr = max(seqNbrs(myID, myID), msg.destSeqNbr);
	      hopCnt = 0;
	      lifetime = AODVparams.ACTIVE_ROUTE_TIMEOUT; 
	    end
	    
	    RREP.type = 2;
	    RREP.hopCnt = hopCnt;
	    RREP.dest = dest;
	    RREP.destSeqNbr = seqNbr;
	    RREP.src = msg.src;
	    RREP.lifetime = lifetime;
	    RREP.intermed = myID;
	    
	    % Send RREP to previous hop
	    ttSendMsg(msg.intermed, RREP, 20);
	    
	    etime = 0.0001;
	    
	  else
	    % We do not have a route to the destination
	    
	    % Rebroadcast RREQ
	    if verbose
	      disp(['Time: ' num2str(noww) ' Node#' num2str(myID) ' sending new broadcast']);
	    end
	    
	    msg.hopCnt = msg.hopCnt + 1;
	    msg.destSeqNbr = max(seqNbrs(myID, msg.dest), msg.destSeqNbr);
	    msg.intermed = myID;
	    
	    ttSendMsg(0, msg, 24);
	    
	    etime = 0.0001;
	    
	  end
	  
	else
	  etime = 0.00001; % used if RREQ is dropped
	end
	
      end
      
     case 2, % RREP
      
      if verbose
	disp(['Node#' num2str(myID) ' got an RREP from Node#' num2str(msg.intermed) ' for destination#' num2str(msg.dest)]);
      end
	
      % Create forward route entry, one could already exist
      dest_entry = findEntry(msg.dest, myID);
      
      if (isempty(dest_entry))
	% No entry exists, create new
	if verbose
	  disp(['Creating new forward entry from Node#' num2str(myID) ' to Node#' num2str(msg.dest)]);
	end
	  
	fwd_entry.dest = msg.dest;
	fwd_entry.nextHop = msg.intermed;
	fwd_entry.hops = msg.hopCnt + 1; % New Hop Count
	fwd_entry.destSeqNbr = msg.destSeqNbr;
	fwd_entry.exptime = ttCurrentTime + msg.lifetime;
	fwd_entry.nbor = [];
	fwd_entry.state = 'valid';
	
	l = length(routing_table{myID});
	routing_table{myID}{l+1} = fwd_entry;
	
	% Update expiry timer	    
	updateExpiryTimer(myID);
	
	propagate = 1;
	
	etime2 = 0.0001;
      else
	% Forward entry already exists in table
	% Should it be updated?
	cond1 = strcmp(dest_entry.state, 'invalid'); 
	cond2 = (strcmp(dest_entry.state, 'valid')) & (msg.destSeqNbr > dest_entry.destSeqNbr);
	cond3 = (msg.destSeqNbr > dest_entry.destSeqNbr) & (msg.hopCnt+1 < dest_entry.hops);
	
	if (cond1 | cond2 | cond3)
	  % Update existing entry
	  if verbose
	    disp(['Updating existing forward entry from Node#' num2str(myID) ' to Node#' num2str(msg.dest)]);
	  end
	  
	  dest_entry.nextHop = msg.intermed;
	  dest_entry.hops = msg.hopCnt + 1; % New Hop Count
	  dest_entry.destSeqNbr = msg.destSeqNbr;
	  dest_entry.exptime = ttCurrentTime + msg.lifetime;
	  if (strcmp(dest_entry.state,'invalid'))
	    dest_entry.nbor = [];
	  end
	  dest_entry.state = 'valid';
	  fwd_entry = dest_entry;
	  
	  updateEntry(dest_entry, myID);

	  propagate = 1;

	  etime2 = 0.0001;
	else
	  % Existing entry should be kept
	  % Do not propagate RREP 
	  if verbose
	    disp(['No entry updated in Node#' num2str(myID)]);
	  end
	    
	  propagate = 0;
		
	  etime2 = 0.0001;
	end
      end
      
      etime1 = 0;
      if (msg.src == myID)
	% Original source, no reverse entry exists
	if verbose
	  disp(['Node#' num2str(myID) ' got final RREP for route to Node#' num2str(msg.dest)]);
	end
	  
	% Start AODVSend to send buffered data messages
	sendMsg.dest = msg.dest;
	ttTryPost('AODVSendBox', sendMsg);
	ttCreateJob('AODVSendTask');
	
	nextHop = [];
	
	etime1 = 0.0001;
	
      elseif (propagate)
		
	% Update reverse entry from info in RREP
	% and get next hop towards source
	nextHop = updateReverse(msg, myID);
	
	% Update precursor list for forward entry
	ind = [];
	if ~isempty(fwd_entry.nbor)
	  ind = find(fwd_entry.nbor==nextHop);
	end
	if (isempty(ind))
	  fwd_entry.nbor = [fwd_entry.nbor nextHop];
	end
	updateEntry(fwd_entry, myID);    
	
	% Update RREP to continue back propagation
	RREP = msg;
	RREP.hopCnt = msg.hopCnt + 1;
	RREP.intermed = myID;
	
	ttSendMsg(nextHop, RREP, 20); 
	
	etime1 = 0.0001;
      
      end
      
      etime = etime1 + etime2;
      
     case 3, % RERR

      if verbose
	disp(['Node#' num2str(myID) ' got an RERR from Node#' num2str(msg.intermed) ' for destination#' num2str(msg.dest)]);
      end
      
      % Propagate RERR?
      
      table = routing_table{myID};
  
      for m = 1:length(table)
	if (strcmp(table{m}.state, 'valid') & table{m}.dest == msg.dest & table{m}.nextHop == msg.intermed)
	  
	  % Should send RERR to all nodes in precursor list (neighbors)
	  for n = 1:length(table{m}.nbor)
	    RERRentry.dest = table{m}.dest;
	    RERRentry.destSeqNbr = table{m}.destSeqNbr;
	    RERRentry.nbor = table{m}.nbor(n);
	    
	    l = length(data.RERRlist);
	    data.RERRlist{l+1} = RERRentry;
	  end
	  
	  % Invalidate route
	  if verbose
	    disp(['Node#' num2str(myID) ' invalidating route to Node#' num2str(table{m}.dest) ' through Node#' num2str(msg.intermed)]);
	  end
	    
	  table{m}.state = 'invalid';
	  
	end
      end
      
      % Update table
      routing_table{myID} = table;
      
      etime = 0.0001;
     case 4, % HELLO
      
      if verbose
	disp(['Time: ' num2str(noww) ' Node#' num2str(myID) ' got a hello message from Node#' num2str(msg.dest)]);
      end
	
      dataHello = ttGetData('HelloTask');
            
      % Update time stamp for last HELLO msg
      found = 0;
      for k = 1:length(dataHello.nbors)
	if (dataHello.nbors{k}.nborID==msg.dest)
	  dataHello.nbors{k}.lastHello = noww;
	  found = 1;
	end
      end
      if (~found)
	nbor.nborID = msg.dest;
	nbor.lastHello = noww;
	l = length(dataHello.nbors);
	dataHello.nbors{l+1} = nbor;
      end
      
      ttSetData('HelloTask', dataHello);
	
      etime = 0.0001;
    end
    
  end
  
  exectime = etime;
  
 case 2,
  % Send all RERRs
  if (~isempty(data.RERRlist))
    data.cnt = data.cnt + 1;
    
    if (data.cnt > length(data.RERRlist))
      
      data.RERRlist = {}; 
      data.cnt = 0;
      exectime = -1;
      
    else
      
      RERRentry = data.RERRlist{data.cnt};
      RERR.type = 3;
      RERR.dest = RERRentry.dest;
      RERR.destSeqNbr = RERRentry.destSeqNbr;
      RERR.intermed = data.nodeID;

      if verbose
	disp(['Node#' num2str(myID) ' sending RERR to Node#' num2str(RERRentry.nbor)]);
      end
	
      ttSendMsg(RERRentry.nbor, RERR, 12); 
    
      ttSleep(0.001);
      exectime = 0;
    
    end
    
  else
    exectime = -1;
  end 
  
 case 3,  
  ttSetNextSegment(2);
  exectime = 0;
  
end

File: examples/RUNES_demo/robot/AODV/AODVrcvcode.cpp
double AODVrcvcode(int seg, void* data) {
  AODVrcv_Data* d = (AODVrcv_Data*) data;
  int myID = d->nodeID;
  
  MailboxMsg *mailboxmsg; 
  GenericNwkMsg *nwkmsg;
  DataMsg *datamsg;  
  RREQMsg *rreqmsg;  
  RREPMsg *rrepmsg; 
  RERRMsg *rerrmsg, *rerrmsg2; 
  HelloMsg *hellomsg; 
  
  int i, dest, seqNbr, hopCnt, cond1, cond2;
  int intermed, drop, cached_RREQID, propagate, nextHop;
  double now, etime1, etime2, lifetime;
  double etime = 0.0001;
  
  RouteEntry *dest_entry, *src_entry;
  DataNode *dn;
  
  switch (seg) {
  case 1:
    nwkmsg = (GenericNwkMsg*) ttGetMsg(ZIGBEENETW);
    now = ttCurrentTime();
    
    ///////////////////////
      // Tmote stuff start //
      ///////////////////////
      if(nwkmsg->type == RANGE_RESPONSE) {
	if (ttCurrentTime() < d->rcv_timeout) {
	  // Get data from rangeResponseMsg and then delete it
	  RangeResponseMsg *rangeResp = (RangeResponseMsg *)nwkmsg->msg;
				
	  double x, y;
	  unsigned int range;
	
	  range = rangeResp->range;
	  x = rangeResp->x;
	  y = rangeResp->y;

	  d->responseCounter++;
	  //		mexPrintf("Response counter = %d \n",d->responseCounter);
	
	  // Remove 145 for 2ms delay
	  // Add 80 for compensation for unknown error
	  double dist;
	  //   	dist = 34400.0 * (double)(range - 145 + 80) / 72280.0; original
	  dist = 34400.0 * (double)(range - 145 + 68) / 72280.0;
	  //		dist = 34400.0 * (double)(range - 145) / 72280.0;
	  //		mexPrintf("Distance measurment from %d = %f \n",rangeResp->nodeID,dist);
	
	  // Put values in data
	  d->node_x = x;
	  d->node_y = y;
	  d->node_dist = dist;
	  // DEBUG
	  d->node_nodeID = rangeResp->nodeID;
	  if (fabs(dist) < 300.0) {
	    // Start task that sends data to mega128
	    ttCreateJob("node_data_to_kalman");
	  }
	}
      } else if (nwkmsg->type == RANGE_REQUEST) {
	d->snd_block = ttCurrentTime() + 0.310;
      } 
      //////////////////////
      // Tmote stuff ends //
      //////////////////////
      else if (nwkmsg->type == DATA) {
        // Data message
        datamsg = (DataMsg*) nwkmsg->msg;
        delete nwkmsg;
        
        if (datamsg->dest == myID) { 
	  // Data message arrived at destination
	  if (VERBOSE) {
	    mexPrintf("Time: %f Data message arrived at Node#%d\n", now, myID);
	  }
      
          // Update expiry timer for route to source
          d->routing_table[datamsg->src - 1].exptime = now + ACTIVE_ROUTE_TIMEOUT;
    
          // Notify application
          ttTryPost("AODVRcvBox", datamsg);
          ttCreateJob("RcvTask");
    
          etime = 0.0001;
        } else {
	  // Forward data message
	  if (VERBOSE) {
	    mexPrintf("Time: %f Node#%d about to forward data to Node#%d Data: %f\n", now, myID, datamsg->dest, datamsg->data);  
	  }
      
          // Update expiry timer for route to source
          d->routing_table[datamsg->src - 1].exptime = now + ACTIVE_ROUTE_TIMEOUT;
    
          // Find route to destination and update expiry timer
          dest_entry = &(d->routing_table[datamsg->dest - 1]);
          dest_entry->exptime = now + ACTIVE_ROUTE_TIMEOUT;
   
          // Forward data message to next hop
          nwkmsg = new GenericNwkMsg;
          nwkmsg->type = DATA;
          nwkmsg->msg = datamsg;
          ttSendMsg(dest_entry->nextHop, nwkmsg, datamsg->size); 
    		
          etime = 0.0001;
        }
        updateExpiryTimer(d->routing_table, d->dataTimer, myID);
      } else {
        // AODV control message (RREQ, RREP, RERR, or HELLO)
        if (VERBOSE) {
	  mexPrintf("Time: %f Node#%d processing AODV message type: %d from Node#%d\n", now, myID, nwkmsg->type, nwkmsg->intermed);
        }
    
        switch (nwkmsg->type) {
        case RREQ:
	  rreqmsg = (RREQMsg*) nwkmsg->msg;
	  intermed = nwkmsg->intermed;
            
	  if (rreqmsg->src == myID) {
	    // Skip RREQs received by the original source	
	
	    etime = 0.00001;
	  } else { 
	    // Have this RREQ already been processed?
	    cached_RREQID = d->cache[rreqmsg->src - 1];
	            
	    drop = 0;
	    if (rreqmsg->RREQID <= cached_RREQID) {
	      // Found in cache, drop redundant RREQ
	      if (VERBOSE) {
		mexPrintf("Time: %f Node#%d dropping redundant RREQ from Node#%d\n", now, myID, intermed);
	      }
	      drop = 1;
	    }
				
	    if (!drop) { 
	      // process RREQ
	      if (VERBOSE) {
		mexPrintf("Time: %f Node#%d caching RREQ with Src: %d RREQID: %d\n", now, myID, rreqmsg->src, rreqmsg->RREQID);
	      }
	  
	      // Enter RREQID in cache
	      d->cache[rreqmsg->src - 1] = rreqmsg->RREQID;
	  
	      // Create or update route entry to source
	      src_entry = &(d->routing_table[rreqmsg->src - 1]);
	  
	      if (!src_entry->valid) {
		// No entry exists or invalid
		src_entry->destSeqNbr = rreqmsg->srcSeqNbr;
		for (i=0; i<NBR_AODV; i++) {
		  src_entry->prec[i] = 0; // empty precursor list
		}
	      } else {
		// Update destination sequence number
		src_entry->destSeqNbr = max(src_entry->destSeqNbr, rreqmsg->srcSeqNbr);
	      }
	      src_entry->dest = rreqmsg->src;
	      src_entry->nextHop = intermed;
	      src_entry->hops = rreqmsg->hopCnt + 1;
	      src_entry->exptime = now + ACTIVE_ROUTE_TIMEOUT;
	      src_entry->valid = 1;
	                
	      updateExpiryTimer(d->routing_table, d->dataTimer, myID);
	                
	      // Check if we have a route to destination
	      dest_entry = &(d->routing_table[rreqmsg->dest - 1]);
	  
	      if (rreqmsg->dest==myID || dest_entry->valid) {
		// We are the destination or we have a route to it
		if (VERBOSE) {
		  mexPrintf("Node#%d has a route to destination#%d\n", myID, rreqmsg->dest);
		}
	    
		if (rreqmsg->dest!=myID) {
		  // I am not the destination, but have a route to it
		  if (VERBOSE) {
		    mexPrintf("Sending first RREP from node with route\n");
		  }
	      
		  dest = dest_entry->dest;
		  seqNbr = dest_entry->destSeqNbr;
		  hopCnt = dest_entry->hops;
		  lifetime = dest_entry->exptime - now;
		  dest_entry->prec[intermed] = 1;
		} else {
		  // I am the destination itself
		  if (VERBOSE) {
		    mexPrintf("Sending first RREP from destination itself\n");
		  }
	      
		  dest = myID;
		  if (d->seqNbrs[myID - 1] + 1 == rreqmsg->srcSeqNbr) {
		    d->seqNbrs[myID - 1]++;
		  }
	                        
		  seqNbr = max(d->seqNbrs[myID - 1], rreqmsg->destSeqNbr);
		  hopCnt = 0;
		  lifetime = ACTIVE_ROUTE_TIMEOUT; 
		}
	    
		// Create RREP
		rrepmsg = new RREPMsg;
		rrepmsg->hopCnt = hopCnt;            
		rrepmsg->dest = dest;
		rrepmsg->destSeqNbr = seqNbr;
		rrepmsg->src = rreqmsg->src;
		rrepmsg->lifetime = lifetime;
                    
		// Send RREP to previous hop
		nwkmsg = new GenericNwkMsg;
		nwkmsg->type = RREP;
		nwkmsg->intermed = myID;
		nwkmsg->msg = rrepmsg;       
		ttSendMsg(intermed, nwkmsg, 20*8);  // 20 bytes
	    
		etime = 0.0001;
	      } else {
		// We do not have a route to the destination
		if (VERBOSE) {
		  mexPrintf("Time: %f Node#%d sending new broadcast\n", now, myID);
		}
	                    
		// Update RREQ
		rreqmsg->hopCnt++;
		rreqmsg->destSeqNbr = max(d->seqNbrs[rreqmsg->dest - 1], rreqmsg->destSeqNbr);
	                    
		// Rebroadcast RREQ
		nwkmsg = new GenericNwkMsg;
		nwkmsg->type = RREQ;
		nwkmsg->intermed = myID;
		nwkmsg->msg = rreqmsg;       
		ttSendMsg(0, nwkmsg, 24*8);  // 24 bytes
	    
		etime = 0.0001;
	      }
	    } else {
	      etime = 0.00001; // used if RREQ is dropped
	    }
	  }
	  break;
            
        case RREP:
	  rrepmsg = (RREPMsg*) nwkmsg->msg;
	  intermed = nwkmsg->intermed;
            
	  if (VERBOSE) {
	    mexPrintf("Node#%d got an RREP from Node#%d for destination#%d\n", myID, intermed, rrepmsg->dest);
	  }
	
	  // Initialize or update forward route entry
	  dest_entry = &(d->routing_table[rrepmsg->dest - 1]);
      
	  if (!dest_entry->valid) {
	    // No valid entry exists, initialize new
	    if (VERBOSE) {
	      mexPrintf("Initializing new forward entry from Node#%d to Node#%d\n", myID, rrepmsg->dest);
	    }
	  
	    dest_entry->dest = rrepmsg->dest;
	    dest_entry->nextHop = intermed;
	    dest_entry->hops = rrepmsg->hopCnt + 1; 
	    dest_entry->destSeqNbr = rrepmsg->destSeqNbr;
	    dest_entry->exptime = now + rrepmsg->lifetime;
	    for (i=0; i<NBR_AODV; i++) {
	      dest_entry->prec[i] = 0; // empty precursor list
	    }
	    dest_entry->valid = 1;
	    propagate = 1;
	
	    etime2 = 0.0001;;
	  } else {
	    // Valid forward entry already exists in table
	    // Should it be updated?
	    cond1 = (rrepmsg->destSeqNbr > dest_entry->destSeqNbr);
	    cond2 = ((rrepmsg->destSeqNbr == dest_entry->destSeqNbr) && (rrepmsg->hopCnt+1 < dest_entry->hops));
	
	    if (cond1 || cond2) {
	      // Update existing entry
	      if (VERBOSE) {
		mexPrintf("Updating existing forward entry from Node#%d to Node#%d\n", myID, rrepmsg->dest);
	      }
	  
	      dest_entry->nextHop = intermed;
	      dest_entry->hops = rrepmsg->hopCnt + 1; 
	      dest_entry->destSeqNbr = rrepmsg->destSeqNbr;
	      dest_entry->exptime = now + rrepmsg->lifetime;

	      propagate = 1;
	      etime2 = 0.0001;
	    } else {
	      // Existing entry should be kept
	      // Do not propagate RREP 
	      if (VERBOSE) {
		mexPrintf("No entry updated in Node#%d\n", myID);
	      }
	    
	      propagate = 0;
		
	      etime2 = 0.0001;;
	    }
	  }
      
	  etime1 = 0;
      
	  if (rrepmsg->src == myID) {
	    // Original source, no reverse entry exists
	    if (VERBOSE) {
	      mexPrintf("Node#%d got final RREP for route to Node#%d\n", myID, rrepmsg->dest); 
	    }
	  
	    // Inform AODVSend to send buffered data messages
	    mailboxmsg = new MailboxMsg;
	    mailboxmsg->type = ROUTE_EST;
	    mailboxmsg->dest = rrepmsg->dest;
	    mailboxmsg->datamsg = NULL;
	    ttTryPost("AODVSndBox", mailboxmsg);
	
	    etime1 = 0.0001;
	
	  } else if (propagate) {
		
	    // Update reverse entry from info in RREP
	    // and get next hop towards source
	    src_entry = &(d->routing_table[rrepmsg->src - 1]);
	    src_entry->exptime = now + rrepmsg->lifetime;
	    src_entry->prec[intermed-1] = 1; // the node that sent the RREP is a precursor towards the src
	    nextHop = src_entry->nextHop;
	            
	    // Update precursor list for forward entry
	    dest_entry->prec[nextHop-1] = 1; // the node that will rcv next RREP is a precursor towards the dest
	
	    // Update RREP to continue back propagation
	    rrepmsg->hopCnt++;
	                   
	    nwkmsg = new GenericNwkMsg;
	    nwkmsg->type = RREP;
	    nwkmsg->intermed = myID;
	    nwkmsg->msg = rrepmsg;   
	            
	    ttSendMsg(nextHop, nwkmsg, 20*8);  // 20 bytes
	
	    etime1 = 0.0001;;
      
	  }
	  updateExpiryTimer(d->routing_table, d->dataTimer, myID);
            
	  etime = etime1 + etime2;
	  break;
            
        case RERR:

	  rerrmsg = (RERRMsg*) nwkmsg->msg;
	  intermed = nwkmsg->intermed;
            
	  if (VERBOSE) {
	    mexPrintf("Node#%d got an RERR from Node#%d for destination#%d\n", myID, intermed, rerrmsg->dest);
	  }
      
	  // Propagate RERR?
	  dest_entry = &(d->routing_table[rerrmsg->dest - 1]);
            
	  if (dest_entry->valid && dest_entry->nextHop == intermed) {
	    // Should send RERR to all nodes in precursor list (neighbors)
	    for (i=0; i<NBR_AODV; i++) {
	      if (dest_entry->prec[i]) {
		// Node i+1 uses this node as next hop towards dest, send RERR
		rerrmsg2 = new RERRMsg;            
		rerrmsg2->dest = dest_entry->dest;
		rerrmsg2->destSeqNbr = dest_entry->destSeqNbr;
		rerrmsg2->receiver = i+1;
	                                
		d->RERRlist->appendNode(new DataNode(rerrmsg2, ""));   
	      }
	    }
	                
	    // Invalidate route
	    if (VERBOSE) {
	      mexPrintf("Node#%d invalidating route to Node#%d through Node#%d\n", myID, rerrmsg->dest, intermed);
	    }
	    
	    dest_entry->valid = 0;
	  }
      
          etime = 0.0001;;
          break;
            
        case HELLO:
	  hellomsg = (HelloMsg*) nwkmsg->msg;
	  intermed = nwkmsg->intermed;
	  //delete nwkmsg;
      
	  if (VERBOSE) {
	    mexPrintf("Time: %f Node#%d got a hello message from Node#%d\n", now, myID, hellomsg->dest); 
	  }
	           
	  // Update time stamp for last HELLO msg
	  d->dataHello->nbors[hellomsg->dest - 1] = 1;
	  d->dataHello->lastHello[hellomsg->dest - 1] = now;
	
	  etime = 0.0001;;
        }
    
      }  
      return etime; 
    
  case 2:
    // Send next RERR, if any
    dn = (DataNode*) d->RERRlist->getFirst(); 
    if (dn != NULL) {
      // Retrieve next RERR in list
      rerrmsg = (RERRMsg*) dn->data;
        
      dn->remove();
      delete dn;

      // Send RERR to receiver
      nwkmsg = new GenericNwkMsg;
      nwkmsg->type = RERR;
      nwkmsg->intermed = myID;
      nwkmsg->msg = rerrmsg;
        
      if (VERBOSE) {
	mexPrintf("Node#%d sending RERR to Node#%d\n", myID, rerrmsg->receiver);
      }
        
      ttSendMsg(rerrmsg->receiver, nwkmsg, 12*8);  // 12 bytes 
      ttSleep(0.001);
      return 0.0001;
    } else {
      return FINISHED;    
    }
    
  case 3:
    ttSetNextSegment(2);
    return 0.0001;
  }
 
  return FINISHED; // to supress compilation warnings
}
